### PR TITLE
Add Q# source code formatter

### PR DESCRIPTION
**Pull request by samarsha**
This adds an experimental Q# source code formatter. See the [README](https://github.com/microsoft/qsharp-compiler/blob/samarsha/formatter/src/Formatter/README.md) for current limitations, instructions on how to use it, and an overview of the design.

### CODE REVIEW

File: src/Formatter/App/App.fsproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <AssemblyName>qsfmt</AssemblyName>
+    <RootNamespace>QsFmt.App</RootNamespace>
```
Comment (samarsha) [inline src/Formatter/App/App.fsproj:6]: Does anyone have any preferences for what to call this? We could keep QsFmt and use the namespace `Microsoft.Quantum.QsFmt`. Or something like QsFormat or QsFormatter? Or a longer option is "Q# Formatter" with namespace `Microsoft.Quantum.QSharp.Formatter`...

Then there is also samarsha/qsfmt#2. :)

**Review by Co0olboi — APPROVED** [review]

File: src/Formatter/Formatter.Tests/Examples.fs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+} else {
+return x + 1;
+}
+}
+}""",
+
+    """namespace Foo {
+    function Bar() : Int {
+        let x = 5;
+        if (x == 5) {
```
Comment (cgranade) [inline src/Formatter/Formatter.Tests/Examples.fs:41]: Minor question, but do we want the formatter to keep parens here? I thought that parens after `if` were generally discouraged now that they're optional?

Comment (samarsha) [reply src/Formatter/Formatter.Tests/Examples.fs:41]: Yes, that would be a good formatting rule to add. The formatter hasn't been taught to remove parentheses yet, and I think this test case was written before QEP 1 was implemented. I would like to delay new formatting rule additions to later PRs though.

**Review by swernli — APPROVED** [review]
Excited to see progress on this!

File: src/Formatter/App/App.fsproj (hunk 1 above)
Comment (samarsha) [reply src/Formatter/App/App.fsproj]: We decided to keep QsFmt.

**Pull request merged**